File: j2kt/jre/java/common/java/lang/Long.java
/*
 *  Licensed to the Apache Software Foundation (ASF) under one or more
 *  contributor license agreements.  See the NOTICE file distributed with
 *  this work for additional information regarding copyright ownership.
 *  The ASF licenses this file to You under the Apache License, Version 2.0
 *  (the "License"); you may not use this file except in compliance with
 *  the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package java.lang;

import static javaemul.internal.KtNativeUtils.ktNative;

import javaemul.internal.annotations.KtNative;
import org.jspecify.nullness.NullMarked;
import org.jspecify.nullness.Nullable;

@KtNative(name = "kotlin.Long", companionName = "java.lang.Long")
@NullMarked
public final class Long extends Number implements Comparable<Long> {

  public static final long MAX_VALUE = 0x7fffffffffffffffL;

  public static final long MIN_VALUE = 0x8000000000000000L;

  public static final Class<Long> TYPE = ktNative();

  public static final int SIZE = 64;

  public static final int BYTES = SIZE / Byte.SIZE;

  public Long(long value) {}

  public Long(String string) throws NumberFormatException {}

  @Override
  public native byte byteValue();

  @Override
  public native int compareTo(Long object);

  public static native int compare(long lhs, long rhs);

  public static native Long decode(String string) throws NumberFormatException;

  @Override
  public native double doubleValue();

  @Override
  public native boolean equals(@Nullable Object o);

  @Override
  public native float floatValue();

  public static native long max(long a, long b);

  public static native long min(long a, long b);

  public static native long sum(long a, long b);

  public static native @Nullable Long getLong(@Nullable String string);

  public static native Long getLong(@Nullable String string, long defaultValue);

  public static native @Nullable Long getLong(@Nullable String string, @Nullable Long defaultValue);

  @Override
  public native int hashCode();

  @Override
  public native int intValue();

  @Override
  public native long longValue();

  public static native long parseLong(String string) throws NumberFormatException;

  public static native long parseLong(String string, int radix) throws NumberFormatException;

  @Override
  public native short shortValue();

  public static native String toBinaryString(long v);

  public static native String toHexString(long v);

  public static native String toOctalString(long v);

  @Override
  public native String toString();

  public static native String toString(long n);

  public static native String toString(long v, int radix);

  public static native String toUnsignedString(long n);

  public static native String toUnsignedString(long v, int radix);

  public static native Long valueOf(String string) throws NumberFormatException;

  public static native Long valueOf(String string, int radix) throws NumberFormatException;

  public static native long highestOneBit(long v);

  public static native long lowestOneBit(long v);

  public static native int numberOfLeadingZeros(long v);

  public static native int numberOfTrailingZeros(long v);

  public static native int bitCount(long v);

  public static native long rotateLeft(long v, int distance);

  public static native long rotateRight(long v, int distance);

  public static native long reverseBytes(long v);

  public static native long reverse(long v);

  public static native int signum(long v);

  public static native Long valueOf(long v);

  public static native int hashCode(long l);
}

File: j2kt/jre/java/native/java/lang/Long.kt
/*
 * Copyright 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package java.lang

import javaemul.internal.decodeNumberString
import kotlin.jvm.javaPrimitiveType
import kotlin.math.sign

/** java.lang.Long static method emulations */
object Long {
  operator fun invoke(l: kotlin.Long): kotlin.Long = l

  operator fun invoke(s: kotlin.String): kotlin.Long = s.toLong()

  const val MAX_VALUE = kotlin.Long.MAX_VALUE

  const val MIN_VALUE = kotlin.Long.MIN_VALUE

  const val SIZE = kotlin.Long.SIZE_BITS

  const val BYTES = kotlin.Long.SIZE_BYTES

  val TYPE: Class<kotlin.Long> = kotlin.Long::class.javaPrimitiveType!!

  fun valueOf(l: kotlin.Long): kotlin.Long = l

  fun valueOf(s: kotlin.String): kotlin.Long = s.toLong()

  fun valueOf(s: kotlin.String, radix: Int): kotlin.Long = s.toLong(radix)

  fun compare(l1: kotlin.Long, l2: kotlin.Long): Int = l1.compareTo(l2)

  fun decode(s: kotlin.String): kotlin.Long {
    val (radix, payload) = decodeNumberString(s)
    return payload.toLong(radix)
  }

  fun toString(l: kotlin.Long): kotlin.String = l.toString()

  fun toString(l: kotlin.Long, radix: Int): kotlin.String = l.toString(radix)

  fun toBinaryString(l: kotlin.Long): kotlin.String = l.toULong().toString(2)

  fun toOctalString(l: kotlin.Long): kotlin.String = l.toULong().toString(8)

  fun toHexString(l: kotlin.Long): kotlin.String = l.toULong().toString(16)

  fun toUnsignedString(l: kotlin.Long): kotlin.String = l.toULong().toString()

  fun toUnsignedString(l: kotlin.Long, radix: Int): kotlin.String = l.toULong().toString(radix)

  fun parseLong(s: kotlin.String): kotlin.Long = s.toLong()

  fun parseLong(s: kotlin.String, radix: Int): kotlin.Long = s.toLong(radix)

  fun rotateLeft(l: kotlin.Long, bitCount: Int): kotlin.Long = l.rotateLeft(bitCount)

  fun rotateRight(l: kotlin.Long, bitCount: Int): kotlin.Long = l.rotateRight(bitCount)

  fun hashCode(l: kotlin.Long): Int = l.hashCode()

  fun numberOfLeadingZeros(l: kotlin.Long): Int = l.countLeadingZeroBits()

  fun numberOfTrailingZeros(l: kotlin.Long): Int = l.countTrailingZeroBits()

  fun highestOneBit(l: kotlin.Long): kotlin.Long = l.takeHighestOneBit()

  fun lowestOneBit(l: kotlin.Long): kotlin.Long = l.takeLowestOneBit()

  fun bitCount(l: kotlin.Long): Int = l.countOneBits()

  fun signum(l: kotlin.Long): Int = l.sign

  fun min(l1: kotlin.Long, l2: kotlin.Long): kotlin.Long = kotlin.math.min(l1, l2)

  fun max(l1: kotlin.Long, l2: kotlin.Long): kotlin.Long = kotlin.math.max(l1, l2)

  fun sum(l1: kotlin.Long, l2: kotlin.Long): kotlin.Long = l1 + l2

  fun reverse(l: kotlin.Long): kotlin.Long =
    (Integer.reverse((l shr 32).toInt()).toUInt().toLong()) or
      (Integer.reverse(l.toInt()).toLong() shl 32)

  fun reverseBytes(l: kotlin.Long): kotlin.Long =
    (Integer.reverseBytes((l shr 32).toInt()).toUInt().toLong()) or
      (Integer.reverseBytes(l.toInt()).toLong() shl 32)
}
